internal/exec/stages/files/units.go

```diff
@@ -19,35 +19,63 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/coreos/ignition/v2/config/shared/errors"
 	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
 	"github.com/coreos/ignition/v2/internal/distro"
 	"github.com/coreos/ignition/v2/internal/exec/util"
+	"github.com/coreos/ignition/v2/internal/systemd"
 )
 
+// Preset holds the information about
+// a given systemd unit.
+type Preset struct {
+	unit            string
+	enabled         bool
+	instantiateable bool
+	instances       []string
+}
+
+// warnOnOldSystemdVersion checks the version of Systemd
+// in a given system and prints a warning if older than 240.
+func (s *stage) warnOnOldSystemdVersion() error {
+	systemdVersion, err := systemd.GetSystemdVersion()
+	if err != nil {
+		return err
+	}
+	if systemdVersion < 240 {
+		if err := s.Logger.Warning("The version of systemd (%q) is less than 240. Enabling/disabling instantiated units may not work. See https://github.com/coreos/ignition/issues/586 for more information.", systemdVersion); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // createUnits creates the units listed under systemd.units.
 func (s *stage) createUnits(config types.Config) error {
-	enabledOneUnit := false
+	presets := make(map[string]*Preset)
 	for _, unit := range config.Systemd.Units {
 		if err := s.writeSystemdUnit(unit, false); err != nil {
 			return err
 		}
 		if unit.Enabled != nil {
+			identifier := "disabled"
 			if *unit.Enabled {
-				if err := s.Logger.LogOp(
-					func() error { return s.EnableUnit(unit) },
-					"enabling unit %q", unit.Name,
-				); err != nil {
+				identifier = "enabled"
+			}
+			if strings.Contains(unit.Name, "@") {
+				unitName, instance, err := parseInstanceUnit(unit)
+				if err != nil {
 					return err
 				}
-			} else {
-				if err := s.Logger.LogOp(
-					func() error { return s.DisableUnit(unit) },
-					"disabling unit %q", unit.Name,
-				); err != nil {
-					return err
+				key := fmt.Sprintf("%s-%s", unitName, identifier)
+				if _, ok := presets[key]; ok {
+					presets[key].instances = append(presets[key].instances, instance)
+				} else {
+					presets[key] = &Preset{unitName, *unit.Enabled, true, []string{instance}}
 				}
+			} else {
+				presets[unit.Name] = &Preset{unit.Name, *unit.Enabled, false, []string{}}
```

Comment:
Probably worth a safety check here that we're not clobbering an existing Preset.

Reply:
Addressed your comment.

Reply:
right. this shouldn't happen because igniton checks at a higher level that duplicate entries don't exist in the config, but maybe worth a check here anyway.